compiler/src/dotty/tools/dotc/core/Names.scala

```diff
@@ -289,7 +290,32 @@ object Names {
 
     override def toString =
       if (length == 0) ""
-      else new String(chrs, start, length)
+      else {
+        if (Config.checkBackendNames) {
+          if (!toStringOK) {
+            println("Backend should not call Name#toString, Name#mangledString should be used instead.")
+            new Error().printStackTrace()
+            assert(false)
+          }
+        }
+        new String(chrs, start, length)
+      }
+
+    private def toStringOK = {
+      val trace = Thread.currentThread.getStackTrace
+      !trace.exists(_.getClassName.endsWith("GenBCode")) ||
+      trace.exists(elem =>
+          List(
+              "mangledString",
```

Comment:
An alternative to defining a whitelist in this way would be to define a new method named something like `unmangledString` that does what `toString` currently do, and have `toString` forward to it. Then we can just change all the code calling `toString` to use `unmangledString` instead, but keep the stacktrace check in `toString`